src/libstd/collections/hash/map.rs

```diff
@@ -1208,6 +1208,34 @@ impl<K, V, S> HashMap<K, V, S>
 
         self.search_mut(k).into_occupied_bucket().map(|bucket| pop_internal(bucket).1)
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all pairs `(k, v)` such that `f(&k,&v)` returns `false`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let xs = [(1, 10), (2, 20), (3, 30), (4, 40), (5, 50), (6, 60)];
+    /// let mut map: HashMap<isize, isize> = xs.iter().cloned().collect();
+    /// map.retain(|&k, _| k % 2 == 0);
+    /// assert_eq!(map.len(), 3);
+    /// ```
+    #[stable(feature = "retain_hash_map", since = "1.17.0")]
+    pub fn retain<F>(&mut self, mut f: F)
+        where F: FnMut(&K, &V) -> bool
+    {
+        let tb_cap = self.table.capacity();
+        let old_table = replace(&mut self.table, RawTable::new(tb_cap));
+
+        for (h, k, v) in old_table.into_iter() {
+            if f(&k, &v) {
+                self.insert_hashed_nocheck(h, k, v);
+            }
```

Comment:
The implementation looks good. It could also use the ordered insertion loop that `resize` is using, since we're iterating in hash order and preserving the order.

Either of those are not in place, and ideally this operation should be in place, I wonder if that can be done efficiently? I'd prefer to have this API even if it is not implemented in place. Since it's only ever removing keys, I don't see any traps (cases where it runs very slowly).